aspnetcore/performance/caching/middleware.md

```diff
@@ -1,15 +1,114 @@
+
+```csharp
+var responseCachingFeature = context.HttpContext.Features.Get<IResponseCachingFeature>();
+if (responseCachingFeature != null)
+{
+    responseCachingFeature.VaryByQueryKeys = new[] { "MyKey" };
+}
+```
+
+## HTTP response caching headers
+Response caching by the middleware is configured via your HTTP response headers. The relevant headers are listed below with notes on how they affect caching.
+
+Header | Details
+--- | --- |
+Authorization | <p>The response is not cached if the header exists.</p>
+Cache-Control | <p>The header must be present and its value must explicitly be marked `public`.</p><p>You can control caching with the following parameters:</p><ul><li>max-age</li><li>max-stale</li><li>min-fresh</li><li>must-revalidate</li><li>no-cache</li><li>no-store</li><li>only-if-cached</li><li>private</li><li>public</li><li>s-maxage</li><li>proxy-revalidate</li></ul>
+Pragma | <p>The response is not cached if the value is set to `no-cache`. This header is overridden by the relevant directives in the `Cache-Control` header if present.</p><p>Considered for backward compatibility with HTTP/1.0.</p>
+Set-Cookie | <p>The response is not cached if the header exists.</p>
+Vary | <p>You can vary the cached response by another header. For example, you can cache responses by encoding by including the `Vary: Accept-Encoding` header, which would cache responses to requests with headers `Accept-Encoding: gzip` and `Accept-Encoding: text/plain` separately. A response with a header value of `*` is never stored.</p>
+Expires | <p>A response deemed stale by this header will not be stored or retrieved unless overridden by other `Cache-Control` headers.</p>
+If-None-Match | <p>The full response will be served from cache if the value is not `*` and the `ETag` of the response doesn't match any of the values provided. Otherwise, a 304 (Not Modified) response will be served.</p>
+If-Modified-Since | <p>If the `If-None-Match` header is not present, a full response will be served from cache if the cached response date is newer than the value provided. Otherwise, a 304 (Not Modified) response will be served.</p>
+Date | <p>When serving from cache, the `Date` header is set by the middleware if it wasn't provided on the original response.</p>
+Content-Length | <p>When serving from cache, the `Content-Length` header is set by the middleware if it wasn't provided on the original response.</p>
+Age | <p>The `Age` header sent in the original response will be ignored. The middleware will compute a new value when serving a cached response.</p>
+
+## Troubleshooting
+If caching behavior is not as you expect, confirm that responses are cacheable and capable of being served from the cache by examining the request's incoming headers and the response's outgoing headers. The conditions by which a response will be cached or served from the cache are listed below.
```

Comment:
In fact, it might be worth mentioning that enabling logging can help when debugging. We log why a response is or is not cached and whether it was retrieved from cache for example. If you are curious, the full list of the messages (obviously too much to include here) is at https://github.com/aspnet/ResponseCaching/blob/dev/src/Microsoft.AspNetCore.ResponseCaching/Internal/LoggerExtensions.cs